client/src/app/UI/expression-editor/expression-text-editor/expression-text-editor.component.ts

```diff
@@ -0,0 +1,954 @@
+    
+    constructor(
+        private _datasetService: DataSetService,
+        private _widgetDataService: WidgetRegionDataService,
+    )
+    {
+    }
+
+    ngOnInit()
+    {
+        // Make a copy of incoming expression, so we don't edit what's there!
+        this._expr = new DataExpression(this.expression.id, this.expression.name, this.expression.expression, this.expression.type, this.expression.comments, this.expression.shared, this.expression.creator, this.expression.createUnixTimeSec, this.expression.modUnixTimeSec, this.expression.tags);
+        this.findVariables();
+    }
+
+    ngAfterViewInit(): void
+    {
+        // Get the codemirror object
+        // Vital: https://stackoverflow.com/questions/62097418/how-i-can-add-the-simple-mode-to-codemirror-angular-8
+        // NOTE: As of angular 13 it seems to return a ZoneAwarePromise so we use then to get the object...
+        this._codeMirror.codeMirrorGlobal.then((cm: any)=>
+        {
+            this.setupCodeMirror(cm.default);
+
+            let cmObj = this._codeMirror.codeMirror;
+
+            this._initAsLua = this.isLua;
+
+            cmObj.setOption("mode", this.isLua ? "lua" : "pixlise");
+            cmObj.setOption("lineNumbers", true);
+            cmObj.setOption("theme", "pixlise");
+            cmObj.setOption("readOnly", !this.allowEdit);
+
+            this.setUpKeyBindings(cmObj);
+
+            // Not sure what codemirror is doing, and why it does it but some ms after creation it has been resetting... we now reset it 2x, once for
+            // reducing flicker for user, 2nd time to ensure anything that changed is overwritten with our settings again
+            let setupFunc = ()=>
+            {
+                this.setupCodeMirrorEventHandlers(cmObj);
+                cmObj.refresh();
+            };
+
+            const source1 = timer(1);
+            const sub1 = source1.subscribe(setupFunc);
+
+            const source2 = timer(100);
+            const sub2 = source2.subscribe(setupFunc);
+        });
+    }
+
+    ngOnDestroy()
+    {
+        this._subs.unsubscribe();
+    }
+
+    get isWindows(): boolean
+    {
+        return navigator.userAgent.search("Windows") !== -1;
+    }
+
+    private setUpKeyBindings(cmObj: any): void
+    {
+        const toggleComment = (cm) =>
+        {
+            cm.toggleComment({
+                indent: true,
+                lineComment: this.isLua ? "--" : "//"
+            });
+        };
+
+        const resetMarks = (cm) =>
+        {
+            this.range = null;
+            let cursor = cm.getCursor();
+            cm.setSelection({line: 0, ch: 0}, {line: 0, ch: 0});
+            cm.setSelection(cursor, cursor);
+        };
+        
+        if(this.isWindows)
+        {
+            cmObj.setOption("extraKeys", {
+                "Ctrl-/": toggleComment,
+                "Ctrl-Enter": (cm) =>
+                {
+                    resetMarks(cm);
```

Comment:
Coming from Visual Studio back in the day, I'd want F7 to run and F5 to debug too :) Just kidding though, don't change it!